myst_nbs/case_studies/reinforcement_learning.myst.md

```diff
@@ -0,0 +1,550 @@
+    \end{cases}
+$$
+
+When an action is chosen and a reward $r = \{0,1\}$ is observed, the estimated value of that action is updated as follows:
+
+$$Q_{a} = Q_{a} + \alpha (r - Q_{a})$$
+
+where $\alpha \in [0, 1]$ is a learning parameter that influences how much the value of an action is shifted towards the observed reward in each trial. Finally, the $Q$ table values are converted into action probabilities via the softmax transformation:
+
+$$ P(a = \text{right}) = \frac{\exp(\beta Q_{\text{right}})}{\exp(\beta Q_{\text{right}}) + \exp(\beta Q_{\text{left}})}$$
+
+where the $\beta \in (0, +\infty)$ parameter determines the level of noise in the agent choices. Larger values will be associated with more deterministic choices and smaller values with increasingly random choices.
+
+***
+
+### Credits
+* The PyMC code was adapted from that of Maria Eckstein ([GitHub](https://github.com/MariaEckstein/SLCN), [PyMC Discourse](https://discourse.pymc.io/t/modeling-reinforcement-learning-of-human-participant-using-pymc3/1735))
+* The MLE code was adapted from that of Robert Wilson and Anne Collins {cite:p}`collinswilson2019` ([GitHub](https://github.com/AnneCollins/TenSimpleRulesModeling))
+
+```{code-cell} ipython3
+:id: QTq-0HMw7dBK
+
+import aesara
+import aesara.tensor as at
+import arviz as az
+import matplotlib.pyplot as plt
+import numpy as np
+import pymc as pm
+import scipy
+import seaborn as sns
+
+az.style.use("arviz-darkgrid")
+```
+
+```{code-cell} ipython3
+:id: YuX-31cMC3u7
+
+sns.set(font_scale=1.5)
+seed = sum(map(ord, "RL_PyMC"))
+rng = np.random.default_rng(seed)
+```
+
+```{code-cell} ipython3
+:id: -BmQhhxi6Ol5
+
+def plot_data(actions, rewards, Qs):
+    plt.figure(figsize=(15, 5), layout="constrained")
+    x = np.arange(len(actions))
+
+    plt.plot(x, Qs[:, 0] - 0.5 + 0, c="C0", lw=3, alpha=0.3)
+    plt.plot(x, Qs[:, 1] - 0.5 + 1, c="C1", lw=3, alpha=0.3)
+
+    s = 50
+    lw = 2
+
+    cond = (actions == 0) & (rewards == 0)
+    plt.scatter(x[cond], actions[cond], s=s, c="None", ec="C0", lw=lw)
+
+    cond = (actions == 0) & (rewards == 1)
+    plt.scatter(x[cond], actions[cond], s=s, c="C0", ec="C0", lw=lw)
+
+    cond = (actions == 1) & (rewards == 0)
+    plt.scatter(x[cond], actions[cond], s=s, c="None", ec="C1", lw=lw)
+
+    cond = (actions == 1) & (rewards == 1)
+    plt.scatter(x[cond], actions[cond], s=s, c="C1", ec="C1", lw=lw)
+
+    plt.scatter(0, 20, c="k", s=s, lw=lw, label="Reward")
+    plt.scatter(0, 20, c="w", ec="k", s=s, lw=lw, label="No reward")
+    plt.plot([0, 1], [20, 20], c="k", lw=3, alpha=0.3, label="Qvalue (centered)")
+
+    plt.yticks([0, 1], ["left", "right"])
+    plt.ylim(-1, 2)
+
+    plt.ylabel("action")
+    plt.xlabel("trial")
+
+    handles, labels = plt.gca().get_legend_handles_labels()
+    order = (1, 2, 0)
+    handles = [handles[idx] for idx in order]
+    labels = [labels[idx] for idx in order]
+
+    plt.legend(handles, labels, fontsize=12, loc=(1.01, 0.27))
```

Comment:
Thanks for these comments! I think is a great addition to the notebook. I will add them.

Reply:
So...I added the plot suggestion in https://github.com/pymc-devs/pymc-examples/pull/410/commits/69acb1f64770f06b1098e9b62114f1780f817b8d but I was not able to use `Line2D` to create the legend as you suggested so for now this is something where I would appreciate some tips.

Reply:
Also, I was not able to hide the plot code cell. AFAIK I need to use `:tags: [hide-cell]` in the myst.md file? but then the pre-commit complains because the notebook and the `myst.md` are not consistent ... can you please give me a tip here :) ?

Reply:
> So...I added the plot suggestion in https://github.com/pymc-devs/pymc-examples/commit/69acb1f64770f06b1098e9b62114f1780f817b8d but I was not able to use Line2D to create the legend as you suggested so for now this is something where I would appreciate some tips.

My bad, the marker+linestyle+color string shortcut as 3rd arg is only available for plot, not for Line2D, it needs to use the keyword arguments only:

```
reward_artist = Line2D([], [], c="k", ls="none", marker="o", ms=s, mew=lw, label="Reward")
no_reward_artist = Line2D([], [], ls="none", marker="o", mfc="w", mec="k", ms=s, mew=lw, label="No reward")
Qvalue_artist = Line2D([], [], c="k", ls="-", lw=3, alpha=0.3, label="Qvalue (centered)")

ax.legend(handles=[no_reward_artist, Qvalue_artist, reward_artist], fontsize=12, loc=(1.01, 0.27));
```
> Also, I was not able to hide the plot code cell. AFAIK I need to use :tags: [hide-cell] in the myst.md file? but then the pre-commit complains because the notebook and the myst.md are not consistent ... can you please give me a tip here :) ?

One option is to edit the myst file (one of the reasons for having both myst and ipynb files is to be able to see and review such metadata, you can read https://github.com/pymc-devs/pymc-examples/pull/403#issuecomment-1193133893 if you are interested in this). But it is also possible to do from jupyter. Here you only have to do it once, so it isn't very critical which option you choose, if you had to update a lot of metadata I'd recommend editing the myst file. Overview of the options below:

* Editing myst. MyST and ipynb are synced, it is not that myst is purely an output of the jupyter notebook one. Therefore, if you have to update a lot of metadata in a notebook, I recommend making all the non-metadata changes from jupyter, then commiting that step. At this point, no format has any metadata so they are both perfectly synced, commit to make sure you don't lose anything. Then open the myst file and edit the metadata there, add the hidden cell tags, the figure captions and alt text... after that, run pre-commit which will sync the changes, this time modifying the ipynb file, and commit. The reason for the save+commit between the two is that if you edit both at the same time between syncing then the sync is broken and the only way to fix it is to lose the changes in one of the files, and also, pre-commit is tricky and runs only on staged changes, so if you run it and tries to sync the staged version it can edit the file and overwrite the unstaged one, which effectively means you lose the unstaged changes as these are the ones that are not tracked anywhere.
* Editing from jupyter. jupyter notebook, lab (and I assume vs code extension for jupyter and the like) have ways of editing the cell tags and metadata... This page explains it in more detail: https://jupyterbook.org/en/stable/content/metadata.html. However, the clicking for each cell and (in the case of metadata) having to write json instead of yaml is quite a pain in the ass, hence my recommendation to use myst if making significant metadata/tag changes.

Reply:
Thank you very much for such detailed explanation! (Is this somewhere in the contributing guidelines?) 🚀 I think https://github.com/pymc-devs/pymc-examples/pull/410/commits/cdaecc8003503bb50887670839ceb06299667d69 solved this!

Also, the plot suggestion worked perfectly!